pkg/codegen/importer/hcl2.go

```diff
@@ -0,0 +1,467 @@
+		if err != nil {
+			return nil, fmt.Errorf("invalid provider reference %v: %w", state.Provider, err)
+		}
+		if !providers.IsDefaultProvider(ref.URN()) {
+			name, ok := names[ref.URN()]
+			if !ok {
+				return nil, fmt.Errorf("no name for provider %v", state.Parent)
+			}
+			resourceOptions = appendResourceOption(resourceOptions, "provider", newVariableReference(name))
+		}
+	}
+	if len(state.Dependencies) != 0 {
+		deps := make([]model.Expression, len(state.Dependencies))
+		for i, d := range state.Dependencies {
+			name, ok := names[d]
+			if !ok {
+				return nil, fmt.Errorf("no name for resource %v", d)
+			}
+			deps[i] = newVariableReference(name)
+		}
+		resourceOptions = appendResourceOption(resourceOptions, "dependsOn", &model.TupleConsExpression{
+			Tokens:      syntax.NewTupleConsTokens(len(deps)),
+			Expressions: deps,
+		})
+	}
+	if state.Protect {
+		resourceOptions = appendResourceOption(resourceOptions, "protect", &model.LiteralValueExpression{
+			Tokens: syntax.NewLiteralValueTokens(cty.True),
+			Value:  cty.True,
+		})
+	}
+	return resourceOptions, nil
+}
+
+// typeRank orders types by their simplicity.
+func typeRank(t schema.Type) int {
+	switch t {
+	case schema.BoolType:
+		return 1
+	case schema.IntType:
+		return 2
+	case schema.NumberType:
+		return 3
+	case schema.StringType:
+		return 4
+	case schema.ArchiveType:
+		return 5
+	case schema.AssetType:
+		return 6
+	case schema.JSONType:
+		return 7
+	case schema.AnyType:
+		return 13
+	default:
+		switch t.(type) {
+		case *schema.TokenType:
+			return 8
+		case *schema.ArrayType:
+			return 9
+		case *schema.MapType:
+			return 10
+		case *schema.ObjectType:
+			return 11
+		case *schema.UnionType:
+			return 12
+		default:
+			return int(math.MaxInt32)
+		}
+	}
+}
+
+// simplerType returns true if T is simpler than U.
+//
+// The first-order ranking is:
+//
+//   bool < int < number < string < archive < asset < json < token < array < map < object < union < any
+//
+// Additional rules apply to composite types of the same kind:
+// - array(T) is simpler than array(U) if T is simpler than U
+// - map(T) is simpler than map(U) if T is simpler than U
+// - object({ ... }) is simpler than object({ ... }) if the former has a greater number of required properties that are
+//   simpler than the latter's required properties
+// - union(...) is simpler than union(...) if the former's simplest element type is simpler than the latter's simplest
+//   element type
+func simplerType(t, u schema.Type) bool {
+	tRank, uRank := typeRank(t), typeRank(u)
+	if tRank < uRank {
+		return true
+	} else if tRank > uRank {
+		return false
+	}
+
+	switch t := t.(type) {
+	case *schema.TokenType:
+		u := u.(*schema.TokenType)
+		if t.UnderlyingType != nil && u.UnderlyingType != nil {
+			return simplerType(t.UnderlyingType, u.UnderlyingType)
+		}
+		return false
+	case *schema.ArrayType:
+		return simplerType(t.ElementType, u.(*schema.ArrayType).ElementType)
```

Comment:
Is it safe to assume that `u` is of type `*schema.ArrayType`?

Reply:
Yes: the only way for us to reach this point is for the types to have the same rank, which means that they must be of the same kind.